flow/connectors/clickhouse/clickhouse.go

```diff
@@ -0,0 +1,128 @@
+package connclickhouse
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log/slog"
+	"strings"
+
+	_ "github.com/ClickHouse/clickhouse-go/v2"
+	_ "github.com/ClickHouse/clickhouse-go/v2/lib/driver"
+
+	//_ "github.com/ClickHouse/clickhouse-go"
+
+	"github.com/PeerDB-io/peer-flow/generated/protos"
+	"github.com/PeerDB-io/peer-flow/shared"
+)
+
+//nolint:stylecheck
+type tableNameComponents struct {
+	schemaIdentifier string
+	tableIdentifier  string
+}
+
+type ClickhouseConnector struct {
+	ctx                context.Context
+	database           *sql.DB
+	tableSchemaMapping map[string]*protos.TableSchema
+	logger             slog.Logger
+}
+
+// creating this to capture array results from clicknhouse.
+type ArrayString []string
+
+func (a *ArrayString) Scan(src interface{}) error {
+	switch v := src.(type) {
+	case string:
+		return json.Unmarshal([]byte(v), a)
+	case []byte:
+		return json.Unmarshal(v, a)
+	default:
+		return errors.New("invalid type")
+	}
+}
+
+type UnchangedToastColumnResult struct {
+	TableName             string
+	UnchangedToastColumns ArrayString
+}
+
+func NewClickhouseConnector(ctx context.Context,
+	clickhouseProtoConfig *protos.ClickhouseConfig,
+) (*ClickhouseConnector, error) {
+	database, err := connect(ctx, clickhouseProtoConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open connection to Clickhouse peer: %w", err)
+	}
+
+	flowName, _ := ctx.Value(shared.FlowNameKey).(string)
+	return &ClickhouseConnector{
+		ctx:      ctx,
+		database: database,
+		//database:           conn,
+		tableSchemaMapping: nil,
+		logger:             *slog.With(slog.String(string(shared.FlowNameKey), flowName)),
+	}, nil
+}
+
+func connect(ctx context.Context, config *protos.ClickhouseConfig) (*sql.DB, error) {
+	dsn := fmt.Sprintf("tcp://%s:%d?username=%s&password=%s", //&database=%s"
```

Comment:
do we need to url escape these? can you try a password with special chars and see.

Reply:
Should we pass the db like so to avoid use db below:

```
conn, err = clickhouse.Open(&clickhouse.Options{
            Addr: []string{"<CLICKHOUSE_SECURE_NATIVE_HOSTNAME>:9440"},
            Auth: clickhouse.Auth{
                Database: "default",
                Username: "default",
                Password: "<DEFAULT_USER_PASSWORD>",
            },
            ClientInfo: clickhouse.ClientInfo{
                Products: []struct {
                    Name    string
                    Version string
                }{
                    {Name: "an-example-go-client", Version: "0.1"},
                },
            },

            Debugf: func(format string, v ...interface{}) {
                fmt.Printf(format, v)
            },
            TLS: &tls.Config{
                InsecureSkipVerify: true,
            },
        })
```

https://clickhouse.com/docs/en/integrations/go

Reply:
this is using the clickhouse library interface. We want to use the database/sql interface. With that too there is an option of passing database in the connection string. But when we do that, the queries on the db do not work. I had tried debugging the issue by adding a proxy(in node js) in between clickhouse and client. Figured out the error was due to the database string passed at a wrong location. So finally I settled on "use db" way to fix the issue.